implementations/rust/ockam/ockam_examples/example_projects/tunnel/src/bin/server.rs

```diff
@@ -0,0 +1,83 @@
+use nanorand::{tls_rng, Rng};
+use ockam::{
+    Context, Entity, Result, Routed, SecureChannels, TcpTransport, TrustEveryonePolicy, Vault,
+    Worker,
+};
+
+struct ConnectionBrokerWorker {
+    tcp: TcpTransport,
+    ports: Vec<u16>,
+}
+
+impl ConnectionBrokerWorker {
+    pub fn new(tcp: TcpTransport) -> Self {
+        Self {
+            tcp,
+            ports: (4000..4100).collect(),
+        }
+    }
+}
+
+#[ockam::worker]
+impl Worker for ConnectionBrokerWorker {
+    type Message = String;
+    type Context = Context;
+
+    async fn handle_message(
+        &mut self,
+        ctx: &mut Self::Context,
+        msg: Routed<Self::Message>,
+    ) -> Result<()> {
+        let return_route = msg.return_route();
+
+        // the incoming traffic from the inlet will be redirected
+        // to this address
+        let inlet_dst = {
+            let mut route = return_route.clone();
+            route.modify().pop_back().append(msg.as_body().as_str());
+            route
+        };
+
+        // find a free port
+        let port = {
+            if self.ports.is_empty() {
+                println!("Ran out of free ports!");
+                return Ok(());
+            }
+
+            let idx = tls_rng().generate_range(0..self.ports.len());
+            self.ports.remove(idx)
+        };
+        let addr = format!("0.0.0.0:{}", port);
+
+        // create the inlet
+        self.tcp.create_inlet(addr, inlet_dst).await?;
+
+        println!("Created new tunnel from port {} to {}", port, msg.body());
+
+        // answer with the port that we are listening to
+        ctx.send(return_route, port).await?;
+
+        Ok(())
+    }
+}
+
+#[ockam::node]
+async fn main(ctx: Context) -> Result<()> {
+    // create a secure listening channel
+    let vault = Vault::create(&ctx)?;
+    let mut me = Entity::create(&ctx, &vault)?;
+    me.create_secure_channel_listener("secure_listener", TrustEveryonePolicy)?;
+
+    // start listening over TCP and start worker
+    let tcp = TcpTransport::create(&ctx).await?;
+    ctx.start_worker(
+        "connection_broker",
+        ConnectionBrokerWorker::new(tcp.clone()),
+    )
+    .await?;
+    tcp.listen("0.0.0.0:8000").await?;
+
+    // wait instead of exiting
```

Comment:
Hmm, I think this comment isn't accurate, but feel free to correct if I'm wrong.

I believe we'll exit when you return `Ok`, and the thing that does the waiting is `tcp.listen`.

Reply:
I took the comment from the Rust guide, which [says something similar](https://github.com/ockam-network/ockam/blob/ffadf0633adcae4d0a061fa322fb17dd15afe6e6/examples/rust/get_started/examples/04-routing-over-transport-responder.rs#L18):

> // Don't call ctx.stop() here so this node runs forever.

Reply:
Ah, my bad!

Reply:
Could you make it more clear that the reason we'll run forever is that we aren't calling `ctx.stop()`? I'm new, but I imagine many of the readers of the example will be new as well.